### ISSUE TITLE
Reduce canvas latency: Set `desynchronized` to true for 2d, WebGL, and WebGL2 canvas context options.

### ISSUE THREAD

**Issue by GCDeveloper** [mrdoob/three.js#16684]
Set `desynchronized: true` as a context attribute/option on the 2d, webgl, and webgl2 canvases.

See https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js `contextAttributes = { ... }`

According to this article https://developers.google.com/web/updates/2019/05/desynchronized this setting reduces canvas latency.

> "The desynchronized hint for canvas.getContext() invokes a different code path that bypasses the usual DOM update mechanism. Instead the hint tells the underlying system to skip as much compositing as it is able and in some cases, the canvas's underlying buffer is sent directly to the screen's display controller. This eliminates the latency that would be caused by using the renderer compositor queue."

It is available for:

'2d' | CanvasRenderingContext2D
'webgl' | WebGLRenderingContext
'webgl2' | WebGL2RenderingContext

**Comment by gkjohnson** [comment]
Maybe it's worth documenting but I wasn't able to see much of a difference in the video example (though I haven't tried the scribbling on a touch screen yet). Would this mean that the canvas won't work with difference CSS effects layered on top of it?

The article mentions that not setting `preserveDrawingBuffer` to true can cause flickering, as well, which probably means that if `desynchronized` is defaulted to true then `preserveDrawingBuffer` should be, as well. But that [has its own performance implications](https://stackoverflow.com/questions/27746091/preservedrawingbuffer-false-is-it-worth-the-effort/40098594).

My first inclination is to say default it to `false` and document it but I'd have to play with it more before forming a solid opinion. A compelling example of the benefits would help, as well.

**Comment by GCDeveloper** [comment]
That is fair. Yes I shouldn't say to just set it to true, instead to experiment with it and you're right to keep it defaulted to false for now.

**Comment by Mugen87** [comment]
I thinks it's okay to make a change for this similar to #16102 and then see how it goes.

**Comment by greggman** [comment]
Based on the description of `desynchronized: true` it's really only useful for drawing programs where the user is drawing a small "brush" under their finger. For an app that clears the screen every frame and redraws all the contents (ie, 99.999% of al three.js app) `desynchronized: true` means you'll get flicker since the browser can draw the canvas at any time, in other words when you've only drawn some of the objects not all of them.

TLDR; **don't use `desynchronized: true`  with three.js**

**Comment by greggman** [comment]
`desynchronized: true` is arguably not compatible with 99.999% of three.js apps

`desynchronized: true` is meant for drawing programs where the app draws a "brush" under the user's finger. In this situation the most you'll get is slight tearing for a moment where the brush is drawn.

For an app/page that clears the canvas and then draws all the contents every frame you'll get flicker since the browser is free to show the canvas at any time, including when you haven't finished redrawing all the objects.

`desynchronized: true` is basically a hint to the browser, "if possible single buffer my canvas".

That also means if your app does not flicker that doesn't mean it won't flicker else where. It might not flicker because your particular browser ignores the hint. It might not flicker because the HTML on your page makes the browser choose to ignore the hint. It might not flicker because your GPU is fast and is able to draw all the objects before the browser decides to draw the canvas. And many other reasons.

Basically, unless you're doing the exact case 

1. preserveDrawingBuffer: true,
2. never clearing the canvas
3. drawing on top of the what's already drawn

you should never use `desynchronized: true`

## TLDR; unless you're making a drawing program that draws directly the the canvas never use `desynchronized: true`

**Comment by mrdoob** [comment]
Thanks for the clarification @greggman 🙏

**Comment by greggman** [comment]
you can't verify as the spec is intentionally ambiguous. it allows single buffering with desynchronized:true

https://github.com/whatwg/html/issues/5466
